tests/quick.rs

```diff
@@ -0,0 +1,200 @@
+extern crate evmap;
+
+use evmap::{ReadHandle, WriteHandle};
+
+extern crate quickcheck;
+#[macro_use(quickcheck)]
+extern crate quickcheck_macros;
+
+use quickcheck::Arbitrary;
+use quickcheck::Gen;
+
+use rand::Rng;
+use std::cmp::{min, Ord};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::{BuildHasher, Hash};
+use std::ops::Deref;
+
+fn set<'a, T: 'a, I>(iter: I) -> HashSet<T>
+where
+    I: IntoIterator<Item = &'a T>,
+    T: Copy + Hash + Eq,
+{
+    iter.into_iter().cloned().collect()
+}
+
+#[quickcheck]
+fn contains(insert: Vec<u32>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+
+    insert.iter().all(|&key| r.get(&key).is_some())
+}
+
+#[quickcheck]
+fn contains_not(insert: Vec<u8>, not: Vec<u8>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+
+    let nots = &set(&not) - &set(&insert);
+    nots.iter().all(|&key| r.get(&key).is_none())
+}
+
+#[quickcheck]
+fn insert_empty(insert: Vec<u8>, remove: Vec<u8>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+    for &key in &remove {
+        w.empty(key);
+    }
+    w.refresh();
+    let elements = &set(&insert) - &set(&remove);
+    r.len() == elements.len()
+        && r.map_into::<_, Vec<()>, _>(|_, _| ()).len() == elements.len()
+        && elements.iter().all(|k| r.get(k).is_some())
+}
+
+use Op::*;
+#[derive(Copy, Clone, Debug)]
+enum Op<K, V> {
```

Comment:
If we wanted to be really fancy so that we could also throw `refresh` into this mix, we'd keep two reference maps to reflect "what readers see" separately from the "true" state of the map. And then `refresh` syncs the "true" state to the "reader" map by overwriting it with a clone.

Reply:
Hmm you might've lost me there 🤔 Are you looking for:
1)
```rust
fn do_ops(ops: &[Op], a: &mut WriteHandle, b: &mut HashMap, c: &mut WriteHandle) {
    for op in ops {
        match *op {
            Add(ref k, ref v) => {
                a.insert(k.clone(), v.clone());
                b.entry(k.clone()).or_insert_with(Vec::new).push(v.clone());
                c.insert(k.clone(), v.clone());
            }
            Remove(ref k) => {
                a.empty(k.clone());
                b.remove(k);
                c.empty(k.clone());
            }
            Refresh() => {
              c.refresh()
           }
        }
    }
}
```
or
2)
```rust
fn do_ops(ops: &[Op], a: &mut WriteHandle, b: &mut HashMap, c: &mut WriteHandle) {
    for op in ops {
        match *op {
            Add(ref k, ref v) => {
                a.insert(k.clone(), v.clone());
                b.entry(k.clone()).or_insert_with(Vec::new).push(v.clone());
                c.insert(k.clone(), v.clone());
                c.refresh();
            }
            Remove(ref k) => {
                a.empty(k.clone());
                b.remove(k);
                c.empty(k.clone());
                c.refresh()
            }
        }
    }
}
```

And then is the test:
```rust 
#[quickcheck]
fn test(ops: Vec<Op<Alphabet, i8>>) -> bool {
    let (r, mut w) = evmap::new();
    let (other_r, mut other_w) = evmap::new();
    let mut reference = HashMap::new();
    do_ops(&ops, &mut w, &mut reference, &mut other_w);

    // assert r != other_r
    w.refresh()
    // assert r == other_r
```
?

Reply:
None of those actually!
```rust
fn do_ops(ops: &[Op], w: &mut WriteHandle, r: &mut ReadHandle, wref: &mut HashMap, rref: &mut HashMap) {
    for op in ops {
        match *op {
            Add(ref k, ref v) => {
                wref.insert(k.clone(), v.clone());
                w.entry(k.clone()).or_insert_with(Vec::new).push(v.clone());
            }
            Remove(ref k) => {
                wref.remove(k);
                w.empty(k.clone());
            }
            Refresh => {
                w.refresh();
                *rref = wref.clone();
            }
        }
    }
}
```
And then in the assertion at the end, assert that `ReadHandle == rref`, and that after a `refresh`, `ReadHandle == wref`.

Reply:
Very clever! Okay that makes lots of sense. I hope one day I make as much sense as you do!

Reply:
And just to check, while I'm adding operations, I can't add a `RemoveValue` operation, can I? Because the arbitrary value probably won't be a value in the bag?

Reply:
Oh, you totally can. Just make sure the value set isn't _too_ large. Something like `u8` or a limited alphabet tends to get you quite far.

Reply:
Oh I didn't even notice we were already using `i8`. Considering the implementation of `Large` that's probably good enough?

Reply:
Added `RemoveValue` - 2da788840458579e4adad4358d455c1d1b6c2786